src/box/relay.cc

```diff
@@ -758,13 +768,16 @@ relay_reader_f(va_list ap)
 static inline void
 relay_send_heartbeat(struct relay *relay)
 {
+	if (relay->is_sending_tx)
+		return;
```

Comment:
> If the heartbeat is skipped here, then a next one is sent only after the timeout.

Not really after the timeout, but after the wal is processed. There is a check for `relay->need_new_vclock_sync` in `relay_send_heartbeat_on_timeout`.

>  In case of a too rapid stream of transactions might wait for quite a long undefined time, right?

This is true. I can change that and make relay check if heartbeat dispatch is needed after each sent row.

```diff
diff --git a/src/box/relay.cc b/src/box/relay.cc
index 8581d520a..99c11681e 100644
--- a/src/box/relay.cc
+++ b/src/box/relay.cc
@@ -1240,7 +1240,10 @@ relay_send_row(struct xstream *stream, struct xrow_header *packet)
                                fiber_yield();
                        }
                }
-               relay->is_sending_tx = !packet->is_commit;
                relay_send(relay, packet);
+               relay->is_sending_tx = !packet->is_commit;
+               if (!relay->is_sending_tx) {
+                       relay_send_heartbeat_on_timeout(relay);
+               }
        }
 }
```

I also moved the `is_sending_tx` assignment below, so that it's updated only upon successful row dispatch.